### PR TITLE
[Master] Fix 14254: Return JsDoc tagName when there is no "@" sign prefix

### LINKED ISSUE #14254: Return Completions for @param and other annotations when in JSDoc
From https://github.com/Microsoft/vscode/issues/21239#event-973597091

**TypeScript Version:**  2.2.1

**Code**

```ts
/**
 * |
 */
```

Send a `completions` request against TSServer at the `|`

**Expected behavior:**
Completions for `@params`, `@argument`, ... are returned.

**Actual behavior:**
No results are returned. There has to be an `@` prefix for the jsdoc constructs to be returned:

```ts
/**
 * @|
 */
```

### PR DESCRIPTION
**Pull request by yuit**
Fix #14254

### CODE REVIEW

File: src/services/completions.ts [hunk 1]
```diff
@@ -826,10 +832,20 @@ namespace ts.Completions {
...
-                isJsDocTagName = true;
+            if (hasDocComment(sourceFile, position)) {
+                // The current position is next to the '@' sign, when no tag name being provided yet.
+                // Provide a full list of tag names
+                if (sourceFile.text.charCodeAt(position - 1) === CharacterCodes.at) {
+                    isJsDocTagName = true;
+                }
+                else {
+                    const lineStart = getLineStartPositionForPosition(position, sourceFile);
+                    shouldAppendAtSignBeforeJsDocTagName = sourceFile.text.substr(lineStart, position).indexOf("@") === -1;
```
Comment (mhegazy) [inline src/services/completions.ts]: should not this check of only `*` or whitespaces? `@param` should not be offered in the middle of the comment.

File: src/services/completions.ts [hunk 2]
```diff
@@ -854,8 +870,8 @@ namespace ts.Completions {
                 }
             }
 
-            if (isJsDocTagName) {
-                return { symbols: undefined, isGlobalCompletion: false, isMemberCompletion: false, isNewIdentifierLocation: false, location: undefined, isRightOfDot: false, isJsDocTagName };
+            if (isJsDocTagName || shouldAppendAtSignBeforeJsDocTagName) {
```
Comment (mhegazy) [inline src/services/completions.ts]: looks like `shouldAppendAtSignBeforeJsDocTagName === true` => `isJsDocTagName  === true `

File: src/services/completions.ts [hunk 3]
```diff
@@ -815,6 +815,12 @@ namespace ts.Completions {
         const isJavaScriptFile = isSourceFileJavaScript(sourceFile);
 
         let isJsDocTagName = false;
+        // This is for the case when users request completion in JsDoc without "@"
+        // i.e.
+        //      /**
+        //       * |completion here|
+        //       **/
+        let shouldAppendAtSignBeforeJsDocTagName = false;
```
Comment (mhegazy) [inline src/services/completions.ts]: nit, consider switching this to `isJSDocTag`, where `@param` is a tag, and `param` is a tag name.

File: src/services/jsDoc.ts [hunk 4]
```diff
@@ -88,15 +89,27 @@ namespace ts.JsDoc {
...
-    export function getAllJsDocCompletionEntries(): CompletionEntry[] {
-        return jsDocCompletionEntries || (jsDocCompletionEntries = ts.map(jsDocTagNames, tagName => {
-            return {
-                name: tagName,
-                kind: ScriptElementKind.keyword,
-                kindModifiers: "",
-                sortText: "0",
-            };
-        }));
+    export function getAllJsDocCompletionEntries(shouldAppendAtSign: boolean): CompletionEntry[] {
```
Comment (mhegazy) [inline src/services/jsDoc.ts]: i would split these into two functions, `getJSDocTagNameCompletions` and `getJSDocTagCompletions` and call them based on the context.

File: src/services/completions.ts (hunk 1 above)
Comment (yuit) [inline src/services/completions.ts]: @mhegazy Good point.  I agree above check will allow tagname to be in the middle of the comment...though check only for "*" and whitespace may not be enough. Consider following example:
```js
/**
 |completion
*/

/**
          |completion|
 */

/**
  *       |completion|
  */
```

File: src/services/completions.ts [hunk 5]
```diff
@@ -54,7 +59,7 @@ namespace ts.Completions {
         }
 
         // Add keywords if this is not a member completion list
-        if (!isMemberCompletion && !isJsDocTagName) {
+        if (!isMemberCompletion && !(requestJsDocTag  && requestJsDocTagName)) {
```
Comment (mhegazy) [inline src/services/completions.ts]: should not this be `!isMemberCompletion && !requestJsDocTag  && !requestJsDocTagName`?

**Pull request merged**